############################################################
# Title: Microbiome diversity and community structure analysis
# Project: Soybean developmental stage microbiome study
#
# Description:
#   This script performs microbiome diversity and community
#   structure analyses using OTU tables generated by DADA2.
#   Analyses include:
#     - Rarefaction curves
#     - Alpha diversity (Observed, Chao1, Shannon, Simpson, etc.)
#     - Statistical testing (Kruskal–Wallis, Conover post-hoc)
#     - Beta diversity (Bray–Curtis, PCoA, PERMANOVA)
#
# Input files:
#   - OTU_counts_table.tsv        (OTU count table)
#   - OTU_tax_table.tsv           (Taxonomic annotation)
#   - sample_information.tsv     (Sample metadata)
#
# Output:
#   - Alpha diversity statistics (CSV)
#   - Pairwise PERMANOVA results (CSV)
#   - Figures (rarefaction, alpha diversity, PCoA)
#
# Author: [Your Name]
# Affiliation: [Your Institution]
# Correspondence: [email]
#
# GitHub repository:
#   https://github.com/USERNAME/REPOSITORY
#
# Date: 2026-XX-XX
############################################################


############################
# 1. Environment setup
############################

# R version >= 4.2.0 recommended
# Install missing packages before running

required_pkgs <- c(
  "dada2", "DECIPHER", "tidyverse", "phyloseq", "tigger",
  "reshape2", "RColorBrewer", "ggpubr", "igraph",
  "ggraph", "tidygraph", "ggrepel", "iNEXT",
  "vegan", "conover.test", "multcompView",
  "pairwiseAdonis", "ggtern"
)

invisible(lapply(required_pkgs, require, character.only = TRUE))


############################
# 2. Data import
############################

# Define input directory (user-modifiable)
data_dir <- "data/"

# OTU count table
otu_table_raw <- read.csv(
  file.path(data_dir, "OTU_counts_table.tsv"),
  sep = "\t", row.names = 1
)

OTU <- otu_table(t(otu_table_raw), taxa_are_rows = FALSE)

# Taxonomy table
tax_table_raw <- read.csv(
  file.path(data_dir, "OTU_tax_table.tsv"),
  sep = "\t", row.names = 1
)

TAX <- tax_table(as.matrix(tax_table_raw))

# Sample metadata
sample_metadata <- read.csv(
  file.path(data_dir, "sample_information.tsv"),
  sep = "\t", row.names = 1
)

SAM <- sample_data(sample_metadata)


############################
# 3. Phyloseq object
############################

physeq <- phyloseq(OTU, TAX, SAM)


############################
# 4. Rarefaction analysis
############################

rarefaction_data <- ggiNEXT(
  iNEXT(as.data.frame(t(otu_table_raw)))
)$data

rarefaction_data$stage <- factor(
  rarefaction_data$stage,
  levels = c(
    "Bud", "Mature_flower", "Flower_half_open",
    "Flower_full_opne", "R2", "R3", "R4", "R5",
    "Young_stem", "Old_stem"
  )
)

# Rarefaction curve plot
ggplot(
  rarefaction_data,
  aes(x = x, y = y, color = tissue)
) +
  geom_line() +
  facet_wrap(tissue ~ stage, scales = "free") +
  theme_light() +
  labs(
    x = "Sequencing depth (reads)",
    y = "Inferred OTU richness"
  )


############################
# 5. Alpha diversity
############################

alpha_div <- plot_richness(physeq)$data

alpha_div <- alpha_div %>%
  filter(variable %in% c(
    "Observed", "Chao1", "ACE",
    "Shannon", "Simpson", "InvSimpson", "Fisher"
  )) %>%
  rename(
    SampleID = name,
    Index = variable,
    Value = value
  ) %>%
  left_join(
    data.frame(sample_metadata, SampleID = rownames(sample_metadata)),
    by = "SampleID"
  )

# Kruskal–Wallis test (example: Shannon index)
kruskal.test(Value ~ stage, data = subset(alpha_div, Index == "Shannon"))


############################
# 6. Beta diversity (PCoA)
############################

bray_dist <- vegdist(
  t(apply(otu_table_raw, 1, function(x) x / sum(x))),
  method = "bray"
)

pcoa_res <- cmdscale(bray_dist, eig = TRUE, k = 2)

pcoa_df <- data.frame(
  PCoA1 = pcoa_res$points[, 1],
  PCoA2 = pcoa_res$points[, 2],
  sample_metadata
)

# PERMANOVA
adonis2(bray_dist ~ stage, data = sample_metadata)

# PCoA plot
ggplot(
  pcoa_df,
  aes(PCoA1, PCoA2, color = stage, shape = tissue)
) +
  geom_point(size = 3, alpha = 0.7) +
  theme_classic() +
  labs(
    x = paste0("PCoA1 (", round(pcoa_res$eig[1] / sum(pcoa_res$eig) * 100, 2), "%)"),
    y = paste0("PCoA2 (", round(pcoa_res$eig[2] / sum(pcoa_res$eig) * 100, 2), "%)")
  )


############################################################
# End of script
############################################################
